src/main/java/javaslang/control/Validation.java

```diff
@@ -0,0 +1,253 @@
+    }
+
+    static <E,T> Validation<E,T> failure(Supplier<E> supplier) {
+        return new Failure<>(supplier.get());
+    }
+
+    boolean isSuccess();
+
+    boolean isFailure();
+
+    Object get();
+
+    void foreach(Consumer<T> f);
+
+    @Override
+    boolean equals(Object o);
+
+    @Override
+    int hashCode();
+
+    @Override
+    String toString();
+
+    default <U> U fold(Function<? super E,? extends U> fail, Function<? super T,? extends U> success) {
+        if(isFailure()) {
+            E v = ((Failure<E,T>) this).get();
+            return fail.apply(v);
+        } else {
+            T v = ((Success<E,T>) this).get();
+            return success.apply(v);
+        }
+    }
+
+    default Validation<T,E> swap() {
+        if(isFailure()) {
+            E v = ((Failure<E,T>) this).get();
+            return Validation.success(v);
+        } else {
+            T v = ((Success<E,T>) this).get();
+            return Validation.failure(v);
+        }
+    }
+
+    @Override
+    default <U> Validation<E,U> map(Function<? super T,? extends U> mapper) {
+        if(isFailure()) {
+            E v = ((Failure<E,T>) this).get();
+            return Validation.failure(v);
+        } else {
+            T v = ((Success<E,T>) this).get();
+            return Validation.success(mapper.apply(v));
+        }
+    }
+
+    default <U,R> Validation<U,R> bimap(Function<? super E,? extends U> fail, Function<? super T,? extends R> success) {
+        if(isFailure()) {
+            E v = ((Failure<E,T>) this).get();
+            return Validation.failure(fail.apply(v));
+        } else {
+            T v = ((Success<E,T>) this).get();
+            return Validation.success(success.apply(v));
+        }
+    }
+
+    default <U> Validation<U,T> leftMap(Function<? super E,? extends U> f) {
+        if(isFailure()) {
+            E v = ((Failure<E,T>) this).get();
+            return Validation.failure(f.apply(v));
+        } else {
+            T v = ((Success<E,T>) this).get();
+            return Validation.success(v);
+        }
+    }
+
+    default <U> Validation<List<E>,U> ap(Validation<List<E>,Function1<T,U>> vv) {
+        if(isSuccess() && vv.isSuccess()) {
+            return success(() -> {
+                Function<T,U> f = ((Success<List<E>,Function1<T,U>>) vv).get();
+                return f.apply(((Success<E,T>)this).get());
+            });
+        } else if(isSuccess() && vv.isFailure()) {
+            return (Failure<List<E>,U>) vv;
+        } else if(isFailure() && vv.isSuccess()) {
+            E e = ((Failure<E,T>) this).get();
+            return failure(List.of(e));
+        } else {
+            return failure(() -> {
+                List<E> e  = ((Failure<List<E>,Function1<T,U>>) vv).get();
+                E  e2 = ((Failure<E,T>) this).get();
+                return e.append(e2);
+            });
+        }
+    }
+
+    default <U> ValidationBuilder<E,T,U> bld(Validation<E,U> v) {
```

Comment:
Thank you for the insight. We need both accumulation styles:

* `v1.combine(v2)...combine(vn)` may be used in fold/reduce.
* The static `Validation.combine(v1, v2, ..., vn)` is a great shortcut. I had this in mind, too.

Do Scalaz/Cats also use a builder? Do you have a link to source-code?

I ask myself if it sufficient if we use tuples or functions of arity 1..8 for accumulated results. If we use validation for a html form for example, the form field will probably exceed the limit of 8. It is not a good idea to blow up the arity of tuples and functions (Scala has 22).

On the other hand it is not possible to use an IndexedSeq instead, holding the accumulated results, because the combined Validations have different Success component types.

[HList](https://github.com/milessabin/shapeless/blob/master/core/src/main/scala/shapeless/ops/hlists.scala) to the rescue?

Example:

```scala
def validate(musician: Musician): StringValidation[Musician] = {
    import scalaz._
    import scalaz.Scalaz._

    def validCurrentBand(band: Option[Band]): StringValidation[Option[Band]] = ???

    def validName(name: String): StringValidation[String] = ???

    def validateAge(born: LocalDate): StringValidation[LocalDate] =
      if (born.isAfter(LocalDate.now().minusYears(12))) "too young".failure
      else born.success

    def validInstrument(instruments: Seq[Instrument]): StringValidation[Seq[Instrument]] = ???

    (validName(musician.name)
      |@| atLeast12(musician.born)
      |@| validInstrument(musician.instruments)
      |@| validCurrentBand(musician.currentBand))(_ => musician)
}
```
(taken from https://www.innoq.com/de/blog/validate-your-domain-in-scala/)